File: dsa-problems/TreesDP/solutions/javascript/TDP-012-kth-ancestor-color-filter.js
const readline = require("readline");
const rl = readline.createInterface({ input: process.stdin, terminal: false });

const lines = [];
rl.on("line", (line) => lines.push(line.trim()));
rl.on("close", () => {
  let idx = 0;
  const n = parseInt(lines[idx++]);
  const LOG = 20;
  const color = [0, ...lines[idx++].split(" ").map(Number)];

  const adj = Array.from({ length: n + 1 }, () => []);
  for (let i = 0; i < n - 1; i++) {
    const [u, v] = lines[idx++].split(" ").map(Number);
    adj[u].push(v);
    adj[v].push(u);
  }

  const up = Array.from({ length: n + 1 }, () => Array(LOG).fill(0));

  function dfs(u, p) {
    up[u][0] = p;
    for (let i = 1; i < LOG; i++) {
      up[u][i] = up[up[u][i - 1]][i - 1];
    }
    for (const v of adj[u]) {
      if (v !== p) dfs(v, u);
    }
  }

  dfs(1, 0);

  function findKth(v, c, k) {
    let count = 0;
    while (v !== 0) {
      if (color[v] === c) {
        count++;
        if (count === k) return v;
      }
      v = up[v][0];
    }
    return -1;
  }

  const q = parseInt(lines[idx++]);
  for (let i = 0; i < q; i++) {
    const [v, c, k] = lines[idx++].split(" ").map(Number);
    console.log(findKth(v, c, k));
  }
});

File: dsa-problems/TreesDP/solutions/python/TDP-012-kth-ancestor-color-filter.py
import sys
sys.setrecursionlimit(300000)

def main():
    data = sys.stdin.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    LOG = 20

    color = [0] + [int(data[idx + i]) for i in range(n)]
    idx += n

    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u, v = int(data[idx]), int(data[idx + 1])
        idx += 2
        adj[u].append(v); adj[v].append(u)

    up = [[0] * LOG for _ in range(n + 1)]

    def dfs(u, p):
        up[u][0] = p
        for i in range(1, LOG):
            up[u][i] = up[up[u][i - 1]][i - 1]
        for v in adj[u]:
            if v != p:
                dfs(v, u)

    dfs(1, 0)

    def find_kth(v, c, k):
        count = 0
        while v != 0:
            if color[v] == c:
                count += 1
                if count == k: return v
            v = up[v][0]
        return -1

    q = int(data[idx]); idx += 1
    for _ in range(q):
        v, c, k = int(data[idx]), int(data[idx + 1]), int(data[idx + 2])
        idx += 3
        print(find_kth(v, c, k))

if __name__ == "__main__":
    main()
